### PR TITLE
Add a commit best practice linked to Deliberate Git

### PR DESCRIPTION
**Pull request by chorn**
I really like Stephen's presentation. A workflow with local WIP commits,
which are squashed into one or more logical commits, with present tense
Subjects and descriptive Bodies is a great practice.

### CODE REVIEW

**Comment by rcdilorenzo** [comment]
@kyptin: I agree. @chorn, it might be good to add a link to this commits markdown file from the main README.

**Comment by gap777** [comment]
I really like what Stephen has to say. I vote for accepting his suggestion as a best practice.

**Comment by gap777** [comment]
@chorn - can you suggest how Stephen's commit message suggestions apply to merge commits? Especially when the feature branch has been squashed to 1 commit, should the merge commit msg be "Add feature XYZ [TR #](GH #)" or more detailed like "Implement XYZView and XYZModel to support the XYZ feature."?

**Comment by chorn** [comment]
I am mostly not a fan of merge commits. :)

My preference is to keep the amount of change in a branch small, and to squash, then rebase onto master without a merge commit at all. I think this is generally called trunk based development. 

For a longer branch, or a branch where the changes made are valuable as separate commits, I prefer to _not_ squash and to `--ff-only` merge.

I want to give future us a chance to climb inside my head and understand what the work at hand was all about while minimizing the cognitive load to interpret the change.

I think that the blend of those two approaches maintains the right balance in the commit history, and maintains our ability to associate change with cards.

@sdball and I have been having a little bit of a discourse on `git log --merges` vs. `git log --no-merges` as far as what we like to look at every day. I'll let you know what he says! :)

**Comment by sdball** [comment]
Aww y'all are so nice!

For my take on making Git history useful at both the high and low level I recommend:

- Keep atomic commits that have been deliberately written to be useful. No need to keep every single commit, but a history rewritten into logical atomic changes is wonderful.
- Use merge commits to serve as the "squash" commit that relates those atomic commits back to the high level feature or fix (i.e. github issue, trello card)
- When a branch is ready to merge use `git rebase master` to ensure a clean history merge and `git merge --no-ff` to ensure a merge commit
- Don't leave the default merge commit message but instead write it to speak to the high level feature work (e.g. the perfect place to put details like "Fixes N" if you're using Github issues
- Use `git log --merges` when you want to see the high level feature story of a repo
- Use `git log --no-merges` when you want to see the linear work story of a repo

With that approach each code line in the repo will still be linked to its atomic change story. IE when you're looking at `set x = 3` you won't see the commit for "Implement Giant Feature" but the lower level commit for "Ensure x never exceeds foo because…"

In pretty pictures, here's how this approach plays out in my [Hwacha](https://github.com/sdball/hwacha) repo.

full_history

just_features

just_atomic

When discussing with @chorn we saw the big drawback with this approach is that you lose the obvious connection between a changed line and its higher level feature. To get that back involves a bit of Git magic that I'm still working out but it _seems_ like this command effectively finds the higher level merge when starting from an atomic commit:

```
# where 4ea8ef7 is an atomic commit
$ git log --oneline  --ancestry-path --merges 4ea8ef7..master | tail -1
* c12c13e  2014-01-05 (2 years, 11 months ago)  Stephen Ball   (tag: v0.2.0) Merge branch 'configuration-block'
```

I'm going to work on wrapping that into something like `git feature` or `git find-merge` or `git part-of` etc.

**Comment by timirwin** [comment]
Thanks for the example @sdball . When you ran `git r --merges`, it seems the messages are the default merge commit messages. Previously, we talked about giving those message something more informative. Is your example just old before making that decision, or do you prefer to just use the default messages? Or something else?

**Comment by kyptin** [comment]
Nice explanation, @sdball, thanks! (The pictures were especially helpful.)

This approach makes sense to me, but I'm interested to hear why we're currently squashing everything as a best practice. I'm wondering if there's some motivation or rationale here that I'm missing, since I wasn't a part of that conversation.

If it turns out there's not a strong reason to squash everything, I think this approach would work out better. I generally like to avoid merge commits when possible, because it makes things like cherry picking and reverting confusing (at least to me). But the benefits that I see are (1) to have a simple criteria to distinguish low-level commits from high-level commits and (2) a mechanism for discussing changes at both levels. Personally, those benefits are enough to overcome my reluctance to use merge commits.

A couple notes about finding the merge commit that an individual commit belongs to: so long as we stick to rebasing, we don't need the `--ancestry-path` option, which excludes commits that don't descend from the commit in question. We shouldn't have any such commits. Also, you can `--reverse` the order and limit the number with `-n` or `--max-count`. So I would propose something like this instead:

    git log --oneline --merges --reverse -n1 ${COMMIT}..master

Or, for use in scripting and git aliases (options like `--oneline` could come after):

    git log --merges --reverse --max-count=1 master --not ${COMMIT}

A couple other useful commands, under this scenario:

1. Find the merge commit(s) given the feature number (assuming we're disciplined about including the card number in the commit message): `git log --merges --grep='\[TR#123\]'`
2. If the merge commit is saved in `${COMMIT}`, find the commits in that merge (including the merge itself): `git log ${COMMIT}~1..${COMMIT}`

**Comment by sdball** [comment]
@timirwin: Yes, the `hwacha` example is only showing the history structure and not the commit message content.

@kyptin: Awesome! Thanks for the improvements to that "find merge commit" command! I'm not sure how your workflow involves cherry-picking so I'm not sure how to address where the problem is with that with the merge commit workflow. For reverts you can revert a merge commit just the same as a single squashed commit and the result is the same. You just also have the option of reverting atomic commits as well if needed.

File: git/commits.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+## [Home](../README.md)
+
+## Commits
```
Comment (rmsbackup) [inline git/commits.md]: Need to link this from the home page

**Pull request closed without merging**